Add Limelight vision subsystem with AprilTag pose estimation

28dchen opened the pull request:
Implemented a complete Limelight-based vision subsystem for robot pose estimation. Specifically, referencing frc1678/C2024-Public

I have created a new branch called limelight. And I've added the Limelight vision subsystem with AprilTag pose estimation in it.

I have added the LimglightConstants configuration class with parameters.
Limelight subsystem that pools pose estimates and can be tuned with Limelight constants.
Update Vision.java to delegate to LimelightSubsystem. It should still have compatibility with the findRobotPose() API
Integrate subsystems in RobotContainer for use by commands/autonomous
Theoretically, the code should estimate the robot position and print out some values to Shuffleboard.

28dchen committed:
Add Limelight vision subsystem with AprilTag pose estimation

Implemented a complete Limelight-based vision subsystem for robot pose estimation. Specifically, referencing frc1678/C2024-Public

28dchen committed:
Update Limelight class to return datas to the shufflebaord

Non-Binary-Programmer reviewed (changes requested):
Looks good overall, just a few comments relating to typing and naming that should be addressed before merge (normally I wouldn't be this picky, but because this is supposed to be a template we should maximize good practices

28dchen committed:
Switching to 2026

28dchen committed:
Revert "Switching to 2026"

This reverts commit bd3463f83c96e47e4bc40e9173f5efaf3dfb2a0b.

28dchen committed:
Switch to 2026

28dchen committed:
Revert "Switch to 2026"

This reverts commit 43aafceee8db15b33a06f98a60d73be4477f7b9d.

28dchen committed:
switch from 2025 to 2026

pi-wasnt-taken committed:
Updated Libraries + Commented Out YAGSL

28dchen committed:
Improve comments and naming

Non-Binary-Programmer committed:
Merged main into this, changing drivebase from yagsl to ctre

Non-Binary-Programmer committed:
fixed things after merge

Non-Binary-Programmer committed:
fixed more things after merge

27mrandazzo commented on build.gradle:
```diff
@@ -146,89 +113,3 @@ wpi.java.configureTestTasks(test)
-}
-
-// Create commit with working changes on event branches
-task(eventDeploy) {
```
Why is this removed? I think we want this. Was this done by the 2026 auto-update, or was this intentional?

27mrandazzo commented on bin/main/frc/robot/Constants$ControllerConstants.class:
I'm not sure why your gradle is placing compiled class files in bin/, and they should be in build/. But regardless, please gitignore these.

27mrandazzo commented on src/main/java/frc/robot/LimelightHelpers.java:
I probably should have filled you in on this, but I added a git submodule located in src/lib/LimelightHelpers whose source is from the git repo itself. This is so we don't have to worry about self-updating. I'd rather you use that one instead of just using the copied LimelightHelpers file in your robot code. Tell me if ya need help or more explanation about this.

27mrandazzo commented on src/main/java/frc/robot/subsystems/LimelightSubsystem.java:
```diff
@@ -0,0 +1,197 @@
+  }
+
+  @SuppressWarnings("PMD.AvoidPrintStackTrace")
+  private String reportObservation(VisionObservation observation) {
```
Uhhhh is this for advantagekit? if so, i don't think this will cut it. Basically, AdvantageKit has its own logging api so a string won't be enough. For now tho, this is OK. Getting better logging for this will be crucial.

27mrandazzo commented on src/main/java/frc/robot/subsystems/Vision.java:
Is there a difference between Vision and Limelight files? It isn't quite clear what the difference is. A good idea might be to have a vision folder/module (call it whatever you want), similar to what we had in Drive.java. Something like:
vision/
vision/Vision.java
And you can put in as many files in here as you want while still being clear about where things belong. Also, it will be more expandable in the future if we improve its logging. But for now I would be happy with just chucking the code into a vision folder as explained above.

27mrandazzo reviewed (changes requested):
Nice work! Just address my comments please. There is a few more things we need (like logging, because AKit uses a special way to log. check out docs or drive subsys) but I think this is a good MVP for vision. 

Of course, we won't really know if it works until someone tests it. Have you tested it and checked how accurate of a vision estimate we can get? If there are any blockers/structural code you need, please let me know.

28dchen committed:
Reorganize package structure and improve logging

28dchen commented:
All fixed! Check [Reorganize package structure and improve logging](https://github.com/athenian-robotics/arc-template/pull/2/commits/5f74a2345a16b81b87e401a82c631f2a0905a1df) commit.

28dchen committed:
Integrate Limelight vision with Swerve Drive Pose Estimator

28dchen committed:
fixed bugs

JS7077 commented on src/main/java/frc/robot/subsystems/vision/Limelight.java:
```diff
@@ -0,0 +1,186 @@
+        && (Timer.getFPGATimestamp() - lastHeartbeatSeconds) <= maxAgeSeconds;
+  }
+
+  /** Convenience wrappers for future commands. */
```
Could you add specific javadocs for the functions and have this comment as a regular comment

JS7077 commented on src/main/java/frc/robot/subsystems/vision/Limelight.java:
```diff
@@ -0,0 +1,186 @@
+        Constants.LimelightConstants.CAMERA_NAME);
+  }
+
+  private boolean isEstimateUsable(PoseEstimate estimate) {
```
While the previous stuff has good explanations, it seems that some of the methods below this lack documentation

JS7077 commented on src/main/java/frc/robot/subsystems/vision/Limelight.java:
```diff
@@ -0,0 +1,186 @@
+    double tagCountFactor = Math.max(1.0, estimate.tagCount);
+    double distanceFactor =
+        estimate.avgTagDist <= 0.0
+            ? 1.0 // Happens if Limelight cannot compute a reliable range from the inputs.
```
The ternary operator part should be deletable, the Math.max() covers it (in my testing)

JS7077 reviewed (commented):
Overall, looks pretty good, for the most part. Could use some more comments in some parts, but mostly really clean. Good job!

28dchen committed:
Add some debug code and update dependencies because it is causing issues

28dchen committed:
Fix AdvantageScope 2D field visualization and vision pose estimation

28dchen committed:
Restructure Vision Subsystem

27mrandazzo committed:
Merge remote-tracking branch 'origin/matty-housekeeping' into limelight

28dchen committed:
Add JavaDoc documentation and improve code clarity in vision subsystem

28dchen committed:
Fix distanceFactor error

Non-Binary-Programmer commented on src/main/java/frc/robot/subsystems/vision/Vision.java:
```diff
@@ -0,0 +1,140 @@
+      return Optional.empty();
+    }
+
+    if (currentPose != null && !measurementMatchesOdometry(currentPose, latestObservation.pose())) {
```
This will cause issues if our odometry gets very far off from the truth, so we should probably have an override.

Non-Binary-Programmer commented on build.gradle:
Build.gradle should stay the same as it is on main right now.

Non-Binary-Programmer reviewed (changes requested):
This is mostly good, just see the comments for some important changes

JS7077 commented on src/main/java/frc/robot/subsystems/vision/Vision.java:
```diff
@@ -28,36 +77,64 @@ public Vision(Limelight limelight) {
+   *
+   * @return true when the raw vision inputs meet the configured thresholds.
+   */
+  private boolean isEstimateUsable(VisionIOInputsAutoLogged inputs) {
```
Could also be written as `return !(first if statement) && !(second if statement)`. (or swap < to >= instead of !)

JS7077 commented on src/main/java/frc/robot/subsystems/vision/VisionIO.java:
```diff
@@ -0,0 +1,35 @@
+
+public interface VisionIO {
+  @AutoLog
+  public static class VisionIOInputs {
```
This should probably be connected to the VisionObservation record in /vision/Vision.java

JS7077 commented on src/main/java/frc/robot/subsystems/vision/VisionIOLimelight.java:
```diff
@@ -0,0 +1,166 @@
+            : Math.max(
+                1.0,
+                estimate.avgTagDist / Constants.LimelightConstants.DISTANCE_TRUST_FALLOFF_METERS);
+        Math.max(1.0, estimate.avgTagDist / Constants.LimelightConstants.DISTANCE_TRUST_FALLOFF_METERS);
```
Line 148 is not being used. I think that it has the same functionality as lines 143-147, so we should probably shorten it down to line 148

pi-wasnt-taken committed:
Updated Documentation + Fixed Overide Issue

pi-wasnt-taken committed:
Swapped > to <= on 137

28dchen committed:
Merge branch 'main' into limelight

28dchen committed:
Merge branch 'limelight' of https://github.com/athenian-robotics/arc-template into limelight

28dchen committed:
idk why limelight lib is been removed and added automatically

27mrandazzo commented on src/main/java/frc/robot/subsystems/drive/Drive.java:
```diff
@@ -155,6 +164,8 @@ public Drive(
             new SysIdRoutine.Mechanism(
                 (voltage) -> runCharacterization(voltage.in(Volts)), null, this));
+
+    SmartDashboard.putData("Field", field);
```
Or, better, put it in an outputs autolog class

27mrandazzo commented on .gitmodules:
I don't think this submodule file should be removed...

27mrandazzo reviewed (dismissed):
There is a _tiny_ bit left. Mainly just revert the build.gradle changes + undelete the build.gradle, plus remove the instances of deprecated logging from SmartDashboard. But other than that, LGTM, I say it is ready to merge!!!!

28dchen committed:
Address code review comments

- Revert build.gradle changes
- Restore .gitmodules
- Remove deprecated SmartDashboard logging from Drive.java

28dchen committed:
Fix build: Add missing maven repositories

28dchen committed:
Fix compilation error in RobotContainer.java

28dchen committed:
Fixed some bugs

Non-Binary-Programmer reviewed (approved):
good for merge!
